text/0000-inline-const.md

```diff
@@ -0,0 +1,322 @@
+before a block.
+
+I'm not aware of equivalents in other languages.
+
+AFAIK, this was [first proposed] by **@scottmcm**.
+
+[zig]: https://kristoff.it/blog/what-is-zig-comptime/#compile-time-function-calls
+[first proposed]: https://internals.rust-lang.org/t/quick-thought-const-blocks/7803/9
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Generic Parameters
+
+Named constants defined inside a function cannot use generic parameters that
+are in scope within that function.
+
+```rust
+fn foo<T>() {
+    const SIZE: usize = { std::mem::size_of::<T>() }; // ERROR
+}
+```
+
+If we use the same rule for inline constants, there would be a class of
+implicitly promotable expressions (e.g. `std::ptr::null::<T>()`, `T::CONST +
+T::ANOTHER_CONST`), that could not be written inside an inline constant. We
+would need to resolve this somehow before we stop promoting these expressions
+as discussed in [future possibilites].
+
+Alternatively, we could allow inline constants to refer to generic parameters.
+Associated constants already behave this way, so I think this is possible.
+However, it may result in more post-monomorphization const-eval errors.
+
+## Naming
+
+I prefer the name inline `const`, since it signals that there is no difference
+between a named `const` and an inline one.
+
+**@scottmcm** prefers "`const` block", which is closer to the syntax and parallels
+the current terminology of `async` block and `unsafe` block. It also avoids any
+accidental conflation with the `#[inline]` attribute, which is unrelated.
+Additionally, it doesn't extend nicely to the single-expression variant
+discussed in [future possibilities].
+
+**@RalfJung** prefers "anonymous `const`". **@scottmcm** mentioned in Zulip
+that this could be confused with the `const _: () = ...;` syntax introduced in
+[RFC 2526]. The reference refers to these as "unnamed" constants.
+
+[RFC 2526]: https://github.com/rust-lang/rfcs/pull/2526
+
+## Lints
+
+As mentioned in the reference-level specification, we need to decide whether we
+want to lint against certain types of inline `const` expressions.
+
+# Future possibilities
+[future possibilities]: #future-possibilities
+
+It would be possible to allow the syntax `const expr` for an inline `const` that
+consists of a single expression. This is analagous to the single expression
+variant of closures: `|| 42`. This is backwards compatible with the current proposal.
+
+Eventually, I would like to try making any expression that could possibly panic
+inelgible for implicit promotion. This includes *all* `const fn` calls as well
+as all arithmetic expressions (e.g., `&(0u32 - 1)`), which currently work due
+to [the way MIR is lowered][arith-assert]. Even though bitwise operators can
+never panic, I would also stop promoting them to be consistent.  This would
+have to be done at an edition boundary. We would only do promotion for
+aggregates, literals, constants and combinations thereof (**@RalfJung** notes
```

Comment:
I just learned from @oli-obk that we also promote `&*` reborrows... so those like should also be shot down then if the goal is to match patterns?

Reply:
I think this is not reflected in the text yet.